dali/kernels/reduce/reduce_setup_utils.h

```diff
@@ -0,0 +1,194 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_REDUCE_REDUCE_SETUP_UTILS_H_
+#define DALI_KERNELS_REDUCE_REDUCE_SETUP_UTILS_H_
+
+#include <utility>
+#include <type_traits>
+#include "dali/core/error_handling.h"
+#include "dali/core/format.h"
+#include "dali/core/small_vector.h"
+#include "dali/core/span.h"
+#include "dali/core/tensor_shape.h"
+#include "dali/core/tensor_shape_print.h"
+#include "dali/core/util.h"
+
+namespace dali {
+namespace kernels {
+namespace reduce_impl {
+
+template <typename T>
+constexpr enable_if_t<std::is_integral<T>::value, int> ilog2(T x) {
+  int n = 0;
+  while (x >>= 1)
+    n++;
+  return n;
+}
+
+
+template <typename OutType = uint64_t, typename BitIndices>
+OutType to_bit_mask(const BitIndices &bit_indices) {
+  OutType mask = 0;
+  for (int idx : bit_indices)
+    mask |= OutType(1) << idx;
+
+  return mask;
+}
+
+/**
+ * @param out_axes a collection of integers, supporting `clear` and `push_back`.
+ * @param dim_groups a collection of integer pairs, supporting `clear` and `push_back`
+ */
+template <typename Axes, typename DimGroups, int ndim>
+void SimplifyReduction(Axes &out_axes, DimGroups &dim_groups,
+                       const TensorListShape<ndim> &in_shape, span<const int> axes) {
```

Comment:
It's not clear, from the documentation, what exactly is a `dim_group`. Could you put a description? (especially, because it's an output parameter)